### ISSUE TITLE
[Windows][3.5] Accessing WinFormsGameWindow.ClientBounds via cross-thread operation crashes

### ISSUE THREAD

**Issue by nilllzz** [MonoGame/MonoGame#4654]
Just upgraded to MonoGame 3.5 and when accessing the ClientBounds property of the WinFormsGameWindow instance "Window" of the Game class from a different thread from the main thread, it crashes with an **InvalidOperationException**.

![error](https://cloud.githubusercontent.com/assets/2119926/13902247/17b5e7c4-ee42-11e5-9e32-2d779eb554c9.PNG)

I tested this with version 3.4, and this behavior is not showing up there.

**Comment by tomspilman** [comment]
The only change there is this one:

https://github.com/mono/MonoGame/commit/e74c7e57227f216a0cba7b50852c72c8bd91e520

I guess the `_form.PointToScreen()` is the culprit, but we would need to debug it to know.

Hard to say if this should be thread safe or not, but if it can be easily made to be thread safe lets do that.

**Comment by nilllzz** [comment]
I tested it, `PointToScreen `seems fine, `ClientSize `appears to be the problem.
But the size is also contained in `_form.ClientRectangle`, so merging those two methods should yield the correct result while still working thread safe.

I'll set up a pull request.

**Comment by nilllzz** [comment]
Hm actually, this might be a little more complex.
Turns out, this only happens in Debug mode in Visual Studio for some reason.
Running the same exe with the same libraries from the bin folder does not produce a crash.

**Comment by theZMan** [comment]
That is expected behavior I believe - the debugger watches for the illegal cross thread operations no debugger = no exception.

What its telling you is that you are performing something that might, one day, on one machine turn out to cause a crash. Likely a crash you will never debug as its on one of your customers machines and will be only useful to earn you a slightly worse steam review ;-)

You need to do UI operations on the main thread see http://stackoverflow.com/questions/1523878/getting-cross-thread-operation-not-valid and many other posts. Seems like something in there counts as a winforms UI operation.

**Comment by nilllzz** [comment]
Yeah, that is the cause of the exception, and the source code that causes this (the `PointToScreen` and `ClientSize` methods) have been identified.

This is changed behavior from MonoGame 3.4 though, and the question is if an exception is intended when accessing `Window.ClientBounds` via a non-GUI thread.
Prior to 3.5, the property did not throw an exception because neither `PointToScreen`, nor `ClientSize` have been used.

**Comment by cra0zy** [comment]
For win forms, accessing GUI from another thread is as easy as: https://github.com/mono/MonoGame/blob/develop/Tools/Pipeline/Windows/MainView.cs#L1032

**Comment by trommlbomml** [comment]
I think this issue can be closed, had a talk with nilllzz. But I think a code annotation for being thread-affine would be nice.

**Comment by nilllzz** [comment]
I think using a dispatcher / invoke is out of question, because it just takes too long for a call that is potentially used very often before MonoGame 3.5.
And I agree, it should at least be indicated in the summary of the Window.ClientBounds property that it cannot be accessed from a non-GUI thread.

**Comment by theZMan** [comment]
As always the question should be 'what does XNA do' in the same circumstance

**Comment by nilllzz** [comment]
XNA does not allow the cross thread operation, so the behavior is the same as MonoGame 3.5.
It also returns the `Window.ClientBounds` with correct X and Y location, not `0, 0`.

Which means that currently, MonoGame perfectly mirrors XNA's behavior.

**Comment by prime31** [comment]
@tomspilman why make this thread safe? All UI related access *should always* be marshaled to the UI thread.

@theZMan XNA is long since dead. There comes a time to move on and that time was 4 years ago. No reason to mimic an ancient API as we move into the Vulkan era.

**Comment by tomspilman** [comment]
> why make this thread safe?

Because it is less error prone to users.  Remember i said "if it can be easily made to be thread safe"...  it doesn't seem like it can be.

> There comes a time to move on

Of course we are fully aware of that.  Here is how we look at things:

 1. If XNA did it really well already... do what XNA did.
 2. If we can do it better, but still work like XNA did... do that.
 3. If what XNA did was crappy..  leave the XNA path in there, but add a better API along with it.
 4. If what XNA did doesn't work anymore at all...  add a new and better API.

This one seems like a 1 or 2 like situation.

**Comment by trommlbomml** [comment]
To clarify: This Issue can be closed. WinForms applications are aware of calls to UI Elements MUST be called in the thread the Form was created. This is not related to MonoGame or XNA behavior.

Making this call always be executed in the main thread is a bad idea. Invoking Methods in different threads - especially when implemented synchronously - is a very bad idea, because this is relatively expensive. For each call you need to check if the function was called from the mainthread. If not, invoke it in the main thread. This is compared to maybe regulary calls very slow. I promise there will be complaints about it.

I think the best solution would to point out it is thread-affine and everything is ok. I think the only workaround would be maybe pinvoking WINAPI, as this may not be thread-affine. But please don't!
